File: android/src/main/java/co/reby/rnkustomersdk/KustomerSDKModule.java
package co.reby.rnkustomersdk;

import android.app.Activity;
import android.support.annotation.NonNull;


import org.json.JSONException;
import org.json.JSONObject;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Promise;

import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.ReadableMapKeySetIterator;
import com.kustomer.ui.Kustomer;

public class KustomerSDKModule extends ReactContextBaseJavaModule {

    private final ReactApplicationContext reactContext;

    public KustomerSDKModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
    }

    @NonNull
    @Override
    public String getName() {
        return "KustomerSDK";
    }

    @ReactMethod
    public void identify(String hash){

        /*Kustomer.identify(hash, new KUSIdentifyListener() {
            @Override
            public void onComplete(final boolean success) {
                // Note: This will be called on background thread
            }
        });*/
    }

    @ReactMethod
    public void presentSupport() {
        final Activity activity = getCurrentActivity();
        //Kustomer.showSupport(activity);

    }

    @ReactMethod
    public void presentKnowledgeBase(){
        final Activity activity = getCurrentActivity();
        //Kustomer.presentKnowledgeBase(activity);
    }

    @ReactMethod
    public void openConversationsCount(final Promise promise){
        //promise.resolve(Kustomer.getOpenConversationsCount());
    }

    @ReactMethod
    public void resetTracking(){

        //Kustomer.resetTracking();
    }

    @ReactMethod
    public void setCurrentPageName(String screen){

        //Kustomer.setCurrentPageName(screen);
    }

    @ReactMethod
    public void describeCustomer(final ReadableMap data){
       /* KUSCustomerDescription customerDescription = new KUSCustomerDescription();

        String email = data.getString("email");
        if(email != null && !email.isEmpty()){
            customerDescription.setEmail(email);
        }

        String phone = data.getString("phone");
        if(phone != null && !phone.isEmpty()){
            customerDescription.setPhone(phone);
        } 

        try {
            ReadableMap customData = data.getMap("custom");
            if (customData != null){
                JSONObject customObject = convertMapToJson(customData);
                customerDescription.setCustom(customObject);
            }
        } catch (JSONException e) {
            e.printStackTrace();
        }

        Kustomer.describeCustomer(customerDescription);*/
    }

    @ReactMethod
    public void customLayout(final ReadableMap data){
    }

    private static JSONObject convertMapToJson(ReadableMap readableMap) throws JSONException {
        JSONObject object = new JSONObject();
        ReadableMapKeySetIterator iterator = readableMap.keySetIterator();
        while (iterator.hasNextKey()) {
            String key = iterator.nextKey();
            switch (readableMap.getType(key)) {
                case Boolean:
                    object.put(key, readableMap.getBoolean(key));
                    break;
                case Number:
                    object.put(key, readableMap.getDouble(key));
                    break;
                case String:
                    String value = readableMap.getString(key);
                    if(value != null && !value.isEmpty()){
                        object.put(key, value);
                    }
                    break;
            }
        }
        return object;
    }
}

File: android/src/main/java/co/reby/rnkustomersdk/KustomerSDKModule.kt
package co.reby.rnkustomersdk

import com.facebook.react.bridge.*
import com.kustomer.core.models.*
import com.kustomer.core.models.chat.*
import org.json.JSONObject
import org.json.JSONException
import com.kustomer.ui.Kustomer
import kotlinx.coroutines.runBlocking

class KustomerSDKModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {
        
    override fun getName(): String {
        return "KustomerSDK"
    }

    @ReactMethod
    fun identify(hash:String, promise: Promise) {
        Kustomer.getInstance().logIn(hash){
            when (it) {
                is KusResult.Success -> promise.resolve(true)
                is KusResult.Error -> it.exception.localizedMessage
            }
        }
    }

    @ReactMethod
    fun presentSupport() {
        Kustomer.getInstance().open()
    }

    @ReactMethod
    fun openConversationsCount(promise: Promise) {
        promise.resolve(Kustomer.getInstance().observeActiveConversationIds())
    }

    @ReactMethod
    fun resetTracking() {
        Kustomer.getInstance().logOut()
    }

    @ReactMethod
    fun describeCustomer(data: ReadableMap) {
        val email = data.getString("email")
        val phone = data.getString("phone")
        val custom = data.getMap("custom")

        val attributes = KusCustomerDescribeAttributes(
                emails = listOf(KusEmail(email!!)),
                phones = listOf(KusPhone(phone!!)),
                custom = toMap(custom!!)
        )

        runBlocking {
            Kustomer.getInstance().describeCustomer(attributes)
        }
    }

    private fun toMap(readableMap: ReadableMap): Map<String, String> {
        val map: MutableMap<String, String> = HashMap()
        val iterator: ReadableMapKeySetIterator = readableMap.keySetIterator()
    
        while (iterator.hasNextKey()) {
            val key: String = iterator.nextKey()
            val type: ReadableType = readableMap.getType(key)

            when (type) {
                ReadableType.Boolean -> map[key] = readableMap.getBoolean(key).toString()
                ReadableType.Number -> map[key] = readableMap.getDouble(key).toString()
                ReadableType.String -> {
                    val value: String? = readableMap.getString(key)
                    if (value != null && !value.isEmpty()) {
                        map[key] = value
                    }   
                }
            }
        }
        return map 
    }
}